Initial API for querying InfluxDB

jsternberg opened the pull request.

jsternberg commented:
I extracted the query interface I had come up with from the writer interface and I'm going to work on them separately. So this PR is for the initial api and implementation for querying the database.

I would like to get some feedback on this and get something merged so that we can start having progress on a new client.

@goller @mark-rushakoff @desa

goller commented:
@jsternberg I'd like to see context.Context used for the queries.

jsternberg commented:
@goller can you give me an example? I'm not entirely sure what that means.

goller commented:
@jsternberg Sure. I think all long or blocking calls should be cancelable.  Here is a blog describing the methodology: https://medium.com/@cep21/how-to-correctly-use-context-context-in-go-1-7-8f2c0fafdf39#.mfai721yt

You don't need to actually use the context's Done at the moment, but, you should have your function signatures take them.

jsternberg commented:
Hm, I'm not sure which function to add that to. The utility functions could certainly use it, but it doesn't really belong in the `Select` call since that should return almost immediately anyway. If you have recommendations of which function to use it in or which function signatures should change,  I'm definitely open to it.

I'll think about the proper place to put it. Maybe the `Select` call should create a context so that it's easier to implement something like Ctrl+C.

Of note, remember that this is an experimental API so even if we merge it, we can change it as much as we want.

mark-rushakoff commented on json_cursor.go:
```diff
@@ -0,0 +1,398 @@
+		return time.Time{}
+	}
+
+	// Attempt to cast this to a string. We return string values for the time.
```
We return string values from time, when the `epoch` query parameter is not provided. I'd speculate keeping the timestamps as integers will perform slightly better for both the server and client. There's probably a benchmark in the standard library to confirm or deny that.

jsternberg replied:
Done. I was not aware of epoch. I'll also have to work in allowing the epoch to be customized, but that might be in a future version.

mark-rushakoff commented on client.go:
```diff
@@ -0,0 +1,214 @@
+// Client is a client that communicates with an InfluxDB server.
+type Client struct {
+	// HTTP client used to talk to the InfluxDB HTTP server.
+	http.Client
```
I'm not sure about embedding the http client, due to it being exported. I know you said elsewhere that you'd like to keep this API experimental, but it would probably be better to keep it unexported so we have less of a breaking change if we decide to swap out the underlying client for something with different performance characteristics - more important for the writer than the querier, probably.

jsternberg replied:
The main reason why I had it this way was so the `Do` function would be available. It might be best to just wrap that one function though. I wanted to expose the mechanics of making a request in case somebody needed to do some form of customizing that, but this may expose too large of a surface area.

I also really wanted to make every struct in `Client` public so that it could be created on the stack if so desired. `NewClient`, while it exists, is more of a convenience function.

jsternberg commented on client.go:
```diff
@@ -0,0 +1,214 @@
+// If this is false and we are using a string for the query, the query will be
+// encoded in the body with the content-type application/x-www-form-urlencoded.
+// If this is true, the query will be encoded in the URL and GET will be used.
+func (c *Client) NewQueryRequest(q interface{}, readonly bool, opt QueryOptions) (*http.Request, error) {
```
I'm not sure the best way to go about removing that boolean parameter. I originally thought about allowing the person to pass `GET` or `POST` to the method like `http.NewRequest`, but it caused more problems than just using a boolean for read-only. That article says to use two methods for two different things, but this method only does one thing even with that parameter. Any suggestions of how to break this function up?

jsternberg replied:
Done.

jsternberg commented:
I just added something new based on influxdata/influxdb#7687. I forgot to add support for bound parameters and I actually forgot we had that feature... Well, either way, I removed `Selectf` and `Executef` because those are just fancy ways of being subject to SQL injection bugs and replaced them with the same pattern [uber-go/zap](https://github.com/uber-go/zap) uses for customizing loggers. There's now an interface `QueryOption` that can be used by us to add customizations to a query. We can extend it for more, but I just added this:

```
querier := influxdb.DefaultQuerier()
querier.Select("SELECT mean(value) FROM cpu WHERE host = $host", influxdb.Param("host", "server01"))
```

The only problem I can see is that bound parameters only work for literals like the above. If you want to use it to substitute for an ident like the `mean(value)` above, it doesn't work. Thoughts?

mark-rushakoff commented:
I think that's reasonable as you have it now. The golang sql package is the same way AFAICT - only placeholder _parameters_ can be used, not arbitrary functions/operators/etc. If the consumer needs to do something different, they can stick with fmt.Sprintf for now.

jsternberg commented:
Other than the `context.Context` stuff, is there anything else that needs to be added before merging this initial API? I'd like to start iterating on this if it's possible.

goller commented:
@jsternberg I really think `context.Context` should be added now.  I think it is really important for network programming where anything can happen.  The `context.Context` gives you the ability to cancel/timeout and such. 

It's now part of the standard library as go 1.7.  However, you don't have to implement it all the way down.  You could use `context.TODO`

jsternberg commented:
I can look into adding `context.Context` in a new PR. This isn't the final API at all and it can be changed.

At the same time, I'm unsure where the context would go. If you could tell me where in the API it should go and how it should be used, I can try adding it.
